enhancements/insights/conditional-data-gathering.md

```diff
@@ -0,0 +1,236 @@
+---
+title: conditional-data-gathering
+authors:
+  - "@Sergey1011010"
+reviewers:
+  - "@inecas"
+  - "@tremes"
+  - "@smarterclayton"
+approvers:
+  - "@smarterclayton"
+creation-date: 2021-07-15
+last-updated: 2021-07-15
+status: implementable
+---
+
+# Conditional Data Gathering
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The conditional gatherer for Insights Operator collects data according to the defined gathering rules*.
+Each rule contains one or more conditions such as "alert A is firing"
+and one or more gatherers with parameters such as "collect X lines of logs from containers in namespace N".
+Current version has these rules defined in the code and the proposal is to load them from an external source
+to make collection of new conditional data faster. It's NOT about running brand new code,
+but just calling the existing gatherers with different validated parameters, so the operator can't
+collect anything we don't expect.
+
+\* note that rule here and later has nothing to do with rules used to analyze data in archives written
+by CCX presentation team
+
+## Motivation
+
+Collecting data on some condition such as an alert is quite common pattern
+during the root-cause analysis and the ability to define the rules
+what to collect on which condition externally will help to do that faster.
+
+### Goals
+
+- Implement validation of the rules for conditional gatherer
+- Create a service taking gathering rules from the git repo and providing them through the API
+- Fetch the rules definitions from the service's API and apply them to the conditional gatherer
+
+### Non-Goals
+
+Empty
+
+## Proposal
+
+Right now we have conditional gatherer which can collect data based on some conditions
+which are defined in the code, for example the next config would collect 100 lines of logs
+from each container in namespace `openshift-cluster-samples-operator`
+and image streams of the same namespace when alert `SamplesImagestreamImportFailing` is firing
+and add it to the next archive.
+
+```json
+[{
+  "conditions": [{
```

Comment:
"conditions" has a rather specific connotation within a kubernetes API. Maybe we can find a different term here? "rules", for example?

Reply:
Well, currently we call conditions + gathering_functions a gathering rule, maybe we need to find better terms cuz it already caused confusing with insights data analysis rules.

Reply:
Yeah, it sounds like "rules" is too overloaded as well.

Reply:
Guys what about "Tentative gathering" the term should hopefully lead people towards the actual fact that this is not happening every time. We won't be able to bypass all the tech lingo though.

Reply:
In the other enhancement (https://github.com/openshift/enhancements/pull/821/files) with which we unified the APIs they call it "cluster condition", maybe we can also use this term since our conditions do pretty much the same thing.

Reply:
Looking at the synonyms for `condition` in dictionaries, `situational`, `contextual` gathering might fit, but seem like they don't offer a good work for `conditions` word here. 

How about `s/conditions/constrains/g`?

Reply:
these are "rules for times/situations in which we should gather these additional pieces of data", right?

conditions seems fine.  rules also seems fine.  constraints does not seem to match.

Reply:
another option if you want to get fancy is `heuristic`

Reply:
Yes, I still feel like `conditions` are still the best fit for what this thinks mean.